src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User actAs(X x, User superUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(superUser.getId());
+
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null ) {
+      throw new AuthenticationException("Super user doesn't exist.");
+    }
+
+    Group group = (Group) groupDAO_.find(user.getGroup());
+
+    if ( group == null && ! group.getEnabled() ) {
+      throw new AuthenticationException("Super user must exist within a group.");
+    }
+
+    /*
+      Finds the AgentUserJunction object to see if user can log in as the passed in user.
+      Source users are permitted to act as target users, not vice versa.
```

Comment:
It should probably be the other way around, right? The person creating the junction is always the source. If Alice wants to act as Bob, then Bob should have to create the junction between himself and Alice, which would give her permission to act as him. In that case, Bob is the source and Alice is the target, so the target is permitted to act as the source, which is the opposite of what's written in the comment. It wouldn't make sense for it to be the other way around from a security perspective because Alice could create the junction between herself and Bob, which would let her act as him. That would mean you could give yourself permission to act as anyone else, which we wouldn't want to allow.